### DISCUSSION TITLE
Layered graph layout

### DISCUSSION THREAD

**Discussion by Adrien-Luxey** [networkx/networkx#4710] [Algorithms]
Hi people!
I'm in the process of making a visualisation of a causally-ordered graph, and I think the [layered graph layout](https://en.wikipedia.org/wiki/Layered_graph_drawing) would best fit my needs.

I just found a very nice course on the Sugiyama method for computing such a layout: [part 1](https://i11www.iti.kit.edu/_media/teaching/winter2016/graphvis/graphvis-ws16-v7-added-proofs.pdf) & [part 2](https://i11www.iti.kit.edu/_media/teaching/winter2016/graphvis/graphvis-ws16-v8.pdf).

I know NetworkX has a [multipartite_layout](https://networkx.org/documentation/stable/reference/generated/networkx.drawing.layout.multipartite_layout.html) function, but apparently it does not handle **node-to-layer assignment**.

Before writing this feature myself, I wanted to know if I maybe missed this piece of code in NetworkX's documentation? Or if you have any example of such an algorithm I could base myself onto?
Also: would NetworkX be interested in such a layout function at all? 

Thank you for your time. NetworkX is a super-duper-library!

**Comment by dschult** [comment]
The highly cited software for this kind of layout is GraphViz and we have two interfaces to that package (pygraphviz and pydot). Assigning the layers is (I believe) an NP-hard problem, though I might be remembering subtle wording that could change the level of hardness.  I'm just trying to say that effective heuristics are key to good performance. Your links seem like a good reference.  We are interested in layout algorithms and including hierarchical layouts like this. But we do have the GraphViz interfaces to enable people to use that tool. Perhaps the most impactful outcome of this effort would be a Python (and hopefully more readable/accessible) version of the method -- which could spur further development.

**Reply by Adrien-Luxey** [reply]
Thank you for your answer @dschult. I will check GraphViz's code, for reference and inspiration. 
Among the four steps of the layered layout process, several are NP-hard indeed! As you saw, the docs I provided thankfully provide (and compare) heuristics. I'm also taking inspiration from NetworkX's clever moves, e.g. [topological_sort](https://networkx.org/documentation/stable/_modules/networkx/algorithms/dag.html#topological_sort) is very efficient.

I'll keep you in touch :)

**Reply by Adrien-Luxey** [reply]
If you know where to locate the algorithm I'm referring to in GraphViz, though, it would help. It's huge and sparsely documented!

**Comment by dschult** [comment]
I believe it is part of the "dot" command line program. That doesn't tell you much.
And I think if it was more clear where that code was or if the heuristics were published 
somewhere , then someone would have already transferred the algorithm somewhere else. :}
Your two links are as good as anything I've seen.  In the last few years a new-ish crew 
of core developers has been working to bring GraphViz code up to better standards. 
But it is open source and hard to get enough time for those willing to do the work.
But the code is improving.  I've mostly worked with the part of the code dealing with 
command line options while helping pygraphviz interface with graphviz. I'm not even
sure what data structure they use for the node/edge information.

**Reply by Adrien-Luxey** [reply]
Interesting stuff seems to be in the [`dotLayout` function](https://gitlab.com/graphviz/graphviz/-/blob/main/lib/dotgen/dotinit.c#L282), but the [companion paper](https://graphviz.org/Documentation/TSE93.pdf) is more readable.
I'll dig this

**Comment by Adrien-Luxey** [comment]
Sorry, I'm talkative, but I have another question: a non-negligible part of the algo is about drawing edges as intelligently placed splines. Is it currently possible to draw splines with [draw_networkx_edges](https://networkx.org/documentation/stable/reference/generated/networkx.drawing.nx_pylab.draw_networkx_edges.html)? Or should I write another function? 

FYI: What the Sugiyama method does is splitting edges that span more than one layer into unit-length edges (separated by properly placed "dummy" vertices). It's not too hard (I guess) to define a spline function for each of these edges as a last step before removing the dummy nodes. The spline parameters could be put in the edges' metadata. However keeping the dummy vertices for drawing is a performance sink: on my test graph, 900% of the vertices are dummy; drawing takes ages.

**Reply by dschult** [reply]
`draw_network_edges` has an argument `connection_style` which lets you draw using matplotlib's arcs, and I believe splines, but I haven't ever used splines so maybe it isn't there.  But the description of what is possible is in `matplotlib.patches.ConnectionStyle`

**Reply by Adrien-Luxey** [reply]
![Sans titre](https://user-images.githubusercontent.com/6992849/113412490-b5e6cf00-93b8-11eb-9b0b-462ae41b388b.png)

Little tease :)

Just missing the last step: vertical placement of nodes so as to minimize edge curvature.

**Comment by Adrien-Luxey** [comment]
Hi,
Let me present you the current results, given a toy graph defined as follows:

```python
G_example = nx.DiGraph()
G_example.add_edges_from([
    (1, 7), (7, 10), (1, 4), (4, 7), (7, 9), (2, 4), (2, 5), (5, 7), (7, 8), (5, 8), (8, 9),
    (2, 6), (6, 8), (3, 5), (3, 6), (3, 9), (9, 10)
])
```
I executed the spring, multipartite, and my layered layout. For the latter, I displayed the graph both with and without the dummy vertices (that participate in node placement and edge drawing - not supported by networkX).

Also note that using the `multipartite_layout` requires to manually annotate vertices with their destination layer as follow, which is not feasible for large graphs:

```python
G_example_mult.nodes[1]['layer'] = 0
G_example_mult.nodes[2]['layer'] = 0
G_example_mult.nodes[3]['layer'] = 0
G_example_mult.nodes[4]['layer'] = 1
G_example_mult.nodes[5]['layer'] = 1
G_example_mult.nodes[6]['layer'] = 1
G_example_mult.nodes[7]['layer'] = 2
G_example_mult.nodes[8]['layer'] = 3
G_example_mult.nodes[9]['layer'] = 4
G_example_mult.nodes[10]['layer'] = 5
```
![example_combined](https://user-images.githubusercontent.com/6992849/113724140-5824f080-96f2-11eb-93b4-532f12c795ee.png)

Now, I didn't implement the whole lot. Missing pieces are:
* Layer width minimization (would really help with large graphs);
* Handling disconnected components (I haven't tried, but chances are that all disconnected components will end up in the first layers, cluttering the display);
* Cycle removal (will allow non-DAG DiGraphs to be laid out);
* Handle scale and center parameters (NetworkX seems to provide functions for this);
* Accept weighted edges - would allow customising the display of MultiDiGraphs (by increasing repeated edges' weight). Not my priority.

Questions regarding incorporation into NetworkX:
* Are you interested in adding this layout to NetworkX? It's more time-consuming than the other layouts, but I feel there is a real added value.
* My code is more than 500 LoC (with lots of comments, but also many subroutines for clarity and DRY). Should I add my code to [networkx/drawing/layout.py](https://github.com/networkx/networkx/blob/ead0e65bda59862e329f2e6f1da47919c6b07ca9/networkx/drawing/layout.py)? It is already more than 1100 LoC... Would a companion file for the layered layout be preferable?
* Drawing edges is an issue. Simply removing dummy edges works well on my example, but it looks awful on bigger graphs. To draw prettier edges, I'd need the ability to draw more complex edges than what is currently provided by `draw_networkx_edges`. Mostly I need the ability to define a path per edge, be it a succession of lines, a spline, or some clever usage of `ConnectionStyle`. Should I try modifying `draw_networkx_edges`?

**Reply by Adrien-Luxey** [reply]
Source code, public domain: https://gist.github.com/Adrien-Luxey/d8f77104056ce5f208a6933467e805f6

**Comment by rossbar** [comment]
Here are my 2 cents regarding your questions:

> Are you interested in adding this layout to NetworkX? It's more time-consuming than the other layouts, but I feel there is a real added value.

Computation time is of course a consideration, but the overriding concern for adding a new layout should be: is it useful? If you believe it has utility that is not covered in other layouts, then yes I think it's worth it. I also second @dschult 's point about the value of having a Python implementation of the method - this is likely to have pedagogical value.

> My code is more than 500 LoC (with lots of comments, but also many subroutines for clarity and DRY). Should I add my code to networkx/drawing/layout.py? It is already more than 1100 LoC... Would a companion file for the layered layout be preferable?

I think the `layout` module is the right place to add a new layout algorithm. It may be near the point where it's worth considering restructuring it as a package, but that bridge can be crossed whenever. Re: LOC - this also is not usually a primary concern - more important is that it is 1) readable and 2) well-written (i.e. idiomatic, makes appropriate use of language features, etc). Fortunately code review can help with these points!

> Drawing edges is an issue.

The edge drawing is (and should remain) decoupled from layout. The general scheme is that layout determines node positions, which also serve as anchor points for edges. The actual drawing is done entirely separately. `nx_draw_edges` is often used for convenience, and as mentioned has parameters like `connection_style` and `arrow_style` to provide a couple simple knobs to tweak edge drawing. If that's not enough for a given use-case, then you may want to explore using matplotlib directly which can offer much more flexibility. Once you've managed to draw the edges in a way that satisfies your use-case, you can then consider if it's a good candidate for upstreaming: is it a good fit for `nx_draw_edges`? Is it a minor addition or will it require a complete rewrite? Of course, you can open a discussion topic on it to get other opinions. If it ends up not being a great fit for `nx_pylab`, another alternative is to create an example for [the gallery](https://networkx.org/documentation/latest/auto_examples/index.html) to highlight e.g. more complex edge drawing.

**Reply by Adrien-Luxey** [reply]
Thank you for your motivating comments :) I do believe that the layered layout would serve some users - at least the ones working on information dissemination like I do. Pretty DAGs are useful DAGs!

Regarding edge drawing, I understand your comments, and will start working on another drawing function dedicated to the layered layout. 
There's only one sad truth: I need to return edges' metadata from the layout function in order to draw them without computing everything twice. It's a shame because all the other layout functions have precisely the same prototype and behaviour. I would need to either return another object in addition to `pos`, or to modify the input graph (adding annotations to edges). I guess the latter is worse, since users don't expect a layout function to modify their input. Would it be acceptable to return `(pos, edges_path)`? 

I'll get working on a PR, thank you again for the motivation!

**Comment by Adrien-Luxey** [comment]
![Sans titre](https://user-images.githubusercontent.com/6992849/114709988-837c9080-9d2d-11eb-9811-956c96fcc673.png)

```python
G = nx.gn_graph(24)
pos, edges_path = nx.layered_layout(G)
nx.draw(G, pos, with_labels=True, edges_path=edges_path)
```
Does this API seem acceptable? `layered_layout` returns two objects instead of one (as the other layout functions do), and a parameter was added to `draw_networkx_edges`. 

Obviously something's wrong with the edge paths, but it kinda works!

(The (6, 1) edge really nails the "be as straight as possible" constraint, doesn't it?)
